pkg/agent/client.go

```diff
@@ -249,11 +252,35 @@ func (a *Client) Connect() (int, error) {
 	}
 	a.conn = conn
 	a.stream = stream
+	a.serverChannel = make(chan *client.Packet, xfrChannelSize)
+	a.serverError = nil
 	a.serverID = serverID
+	go a.writeToKonnServer()
 	klog.V(2).InfoS("Connect to", "server", serverID)
 	return serverCount, nil
 }
 
+func (a *Client) writeToKonnServer() {
+	defer func() {
+		if panicInfo := recover(); panicInfo != nil {
+			klog.V(2).InfoS("Exiting writeToKonnServer with recovery", "panicInfo", panicInfo, "serverID", a.serverID)
+		} else {
+			klog.V(2).InfoS("Exiting writeToKonnServer", "serverID", a.serverID)
+		}
+	}()
+
+	for pkt := range a.serverChannel {
+		klog.V(5).InfoS("writeToKonnServer recevied packet to send to KonnServer", "serverID", a.serverID)
+		err := a.stream.Send(pkt)
+		if err != nil && err != io.EOF {
+			metrics.Metrics.ObserveFailure(metrics.DirectionToServer)
+			a.cs.RemoveClient(a.serverID)
+			klog.V(2).InfoS("WriteToKonnServer removing server from clientset", "serverID", a.serverID)
+			a.serverError = err
```

Comment:
Do we need to lock when setting a.serverError?

Reply:
No. Checking the value of serverError and the setting of serverError to err, all occur on the same goroutine. Adding a comment to that effect.